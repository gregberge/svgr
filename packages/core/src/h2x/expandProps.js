import { JSXAttribute } from 'h2x-plugin-jsx'

const expandProps = (place = 'end') => () => ({
  visitor: {
    JSXElement: {
      enter(path) {
        if (
          path.node.name === 'svg' &&
          !path.node.attributes.some(attr => attr && attr.name === 'props')
        ) {
          const props = new JSXAttribute()
          props.name = 'props'
          props.spread = true
          if (place === 'start') {
            path.node.attributes.unshift(props)
          }
          if (place === 'end') {
            path.node.attributes.push(props)
          }
          path.replace(path.node)
        }
      },
    },
  },
})

export default expandProps
