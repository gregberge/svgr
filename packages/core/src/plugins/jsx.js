import unified from 'unified'
import parse from 'rehype-parse'
import vfile from 'vfile'
import hastToBabelAst from 'hast-util-to-babel-ast'
import { transformFromAstSync, createConfigItem } from '@babel/core'
import removeJSXAttribute from '../babelPlugins/removeJSXAttribute'
import removeJSXEmptyExpression from '../babelPlugins/removeJSXEmptyExpression'

export default (code, config, state = {}) => {
  const filePath = state.filePath || 'unknown'
  const hastTree = unified()
    .use(parse, {
      fragment: true,
      space: 'svg',
      emitParseErrors: true,
      duplicateAttribute: false,
    })
    .parse(vfile({ path: filePath, contents: code }))

  const babelTree = hastToBabelAst(hastTree)

  const { code: transformedCode } = transformFromAstSync(babelTree, code, {
    plugins: [
      createConfigItem([
        removeJSXAttribute,
        { attribute: 'xmlns' },
        'removeXmlns',
      ]),
      createConfigItem(removeJSXEmptyExpression),
    ],
    filename: filePath,
    babelrc: false,
    configFile: false,
  })

  return transformedCode
}
