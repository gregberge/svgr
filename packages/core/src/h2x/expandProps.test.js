import jsx from 'h2x-plugin-jsx'
import { transform } from 'h2x-core'
import expandProps from './expandProps'

describe('expandProps', () => {
  it('should expand props in the end by default', () => {
    const result = transform(
      `<svg width="10" height="10">
        <g stroke="#063855" stroke-width="2" />
      </svg>`,
      { plugins: [jsx, expandProps()] },
    )

    expect(result).toMatchSnapshot()
  })

  it('should expand props in the end with option', () => {
    const result = transform(
      `<svg width="10" height="10">
        <g stroke="#063855" stroke-width="2" />
      </svg>`,
      { plugins: [jsx, expandProps('end')] },
    )

    expect(result).toMatchSnapshot()
  })

  it('should expand props at the start with option', () => {
    const result = transform(
      `<svg width="10" height="10">
        <g stroke="#063855" stroke-width="2" />
      </svg>`,
      { plugins: [jsx, expandProps('start')] },
    )

    expect(result).toMatchSnapshot()
  })
})
