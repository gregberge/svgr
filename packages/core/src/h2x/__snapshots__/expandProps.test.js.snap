// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expandProps should expand props at the start with option 1`] = `
"<svg {...props} width={10} height={10}>
  <g stroke=\\"#063855\\" strokeWidth={2} />
</svg>
"
`;

exports[`expandProps should expand props in the end by default 1`] = `
"<svg width={10} height={10} {...props}>
  <g stroke=\\"#063855\\" strokeWidth={2} />
</svg>
"
`;

exports[`expandProps should expand props in the end with option 1`] = `
"<svg width={10} height={10} {...props}>
  <g stroke=\\"#063855\\" strokeWidth={2} />
</svg>
"
`;
